src/coreclr/src/pal/src/map/virtual.cpp

```diff
@@ -1760,6 +1760,31 @@ VirtualProtect(
     return bRetVal;
 }
 
+#if defined(TARGET_OSX) && defined(TARGET_ARM64)
+bool
+PALAPI
+PAL_JITWriteEnable(bool writeEnable)
+{
+    // Use a thread local to track per thread JIT Write enable state
+    // Threads start with MAP_JIT pages readable and executable (R-X) by default.
+    thread_local bool enabled = false;
```

Comment:
I designed this because I was doing a quick prototype.  I was trying to enable write on non-leaf functions.  I wanted to make nesting work so that if different code generation scenarios had different call stacks, fixing one wouldn't break another.

I am missing why you say nesting wouldn't work....  I can see nesting with exceptions wouldn't work...

+ For best security, we probably should keep the duration of the code writing to a minimum. (Toggle in low level code writing functions.)
+ For best performance, we should minimize these toggles (toggle once per code writing event, currently higher in the stack)
+ To achieve both will require code refactoring.  (Probably more consistent use of something like a code manager)

Reply:
I've said that nesting doesn't work because if you nest two calls as follows, the 3rd call to `PAL_JITWriteEnable` will turn the writability off instead of the 4th one as I would expect with nesting taken into account:
```
PAL_JITWriteEnable(true);
    PAL_JITWriteEnable(true);
    PAL_JITWriteEnable(false);
PAL_JITWriteEnable(false);
```

Reply:
The intended nesting design really was
```C++
bool ov = PAL_JITWriteEnable(true);
    bool nov = PAL_JITWriteEnable(true);
    PAL_JITWriteEnable(nov);
PAL_JITWriteEnable(ov);
```

Using holder approach it still works (but much cleaner)
```C++
auto writeEnableHolder = PAL_JITWriteEnable(true);
    auto nestedWriteEnableHolder = PAL_JITWriteEnable(true);
```